database/apikeys.go

```diff
@@ -150,3 +231,86 @@ func (db *DB) APIKeyList(ctx context.Context, user User) ([]APIKeyRecord, error)
 	}
 	return aks, nil
 }
+
+// APIKeyUpdate updates an existing API key. This works by replacing the
+// list of Skylinks within the API key record. Only valid for public API keys.
+func (db *DB) APIKeyUpdate(ctx context.Context, user User, akID primitive.ObjectID, skylinks []string) error {
+	if user.ID.IsZero() {
+		return errors.New("invalid user")
+	}
+	// Validate all given skylinks.
+	for _, s := range skylinks {
+		if !ValidSkylinkHash(s) {
+			return ErrInvalidSkylink
+		}
+	}
+	filter := bson.M{
+		"_id":     akID,
+		"public":  &True, // you can only update public API keys
+		"user_id": user.ID,
+	}
+	update := bson.M{"$set": bson.M{"skylinks": skylinks}}
+	opts := options.UpdateOptions{
+		Upsert: &False,
+	}
+	ur, err := db.staticAPIKeys.UpdateOne(ctx, filter, update, &opts)
+	if err != nil {
+		return err
+	}
+	if ur.ModifiedCount == 0 {
+		return errors.New("public API key not found, no keys updated")
+	}
+	return nil
+}
+
+// APIKeyPatch updates an existing API key. This works by adding and removing
+// skylinks to its record. Only valid for public API keys.
+func (db *DB) APIKeyPatch(ctx context.Context, user User, akID primitive.ObjectID, addSkylinks, removeSkylinks []string) error {
+	if user.ID.IsZero() {
+		return errors.New("invalid user")
+	}
+	// Validate all given skylinks.
+	for _, s := range append(addSkylinks, removeSkylinks...) {
+		if !ValidSkylinkHash(s) {
+			return ErrInvalidSkylink
+		}
+	}
+	filter := bson.M{
+		"_id":    akID,
+		"public": &True, // you can only update public API keys
+	}
+	var update bson.M
+	// First, all new skylinks to the record.
+	if len(addSkylinks) > 0 {
```

Comment:
this is not great, I think you want to dedupe this first and then also you want to use `$addToSet` instead of `$push`

Reply:
yeah  feel free to resolve, the `$addToSet` was the important one, the reason I would dedupe is that you just make it easier on your database engine 🤷‍♂️